### PR TITLE
Average before producing graphical output.

### PR DESCRIPTION
**Pull request by bangerth**
In reference to #4064. Gets us 90% of the way there. Will probably break a lot of tests -- let's see what happens :-)

/rebuild

### CODE REVIEW

**Comment by bangerth** [comment]
Oh, that's a surprise: No failures?

**Comment by bangerth** [comment]
I've confirmed that none of the tests that do averaging actually output graphical output as well.

**Review by gassmoeller — COMMENTED** [review]
Looks good to me, but have you checked that the output looks reasonable? Maybe add the graphical output to one of the tests that use material averaging, that way we would also have a test for this.

**Comment by bangerth** [comment]
I added a test that starts from the SolKz model, which has an exponentially growing viscosity in z-direction. When visualized without material averaging, one actually sees the exponential growth, but when using material averaging, it shows a stepwise behavior.

This test is replicated three times for arithmetic, geometric, and harmonic averaging. Actually, that's really only what I wanted -- for geometric averaging, the solver doesn't converge, so I didn't press the issue and just check the arithmetic and harmonic averaging, without investigating too deeply what is happening for the geometric case.  I verified that the outputs for the two tests I have look reasonable in comparison to each other, and to the original smooth viscosity.

File: source/postprocess/visualization/material_properties.cc [hunk 1]
```diff
@@ -106,6 +112,36 @@ namespace aspect
...
+        // between input and output of the averaging operation is generally
+        // small and probably not visible anyway.
+        //
+        // The average() function checks whether any quadrature object it uses
+        // has the correct size. So passing an invalid object carries little
+        // risk if the list of averaging modes that require a quadrature
+        // object expands: Every time we generate graphical output for
+        // a model that uses this kind of averaging, we will trigger an
+        // exception.
+        Quadrature<dim> quadrature;
```
Comment (tjhei) [inline source/postprocess/visualization/material_properties.cc:131]: is this variable used? It seems you construct an unnamed quadrature in the call below.

Comment (bangerth) [reply source/postprocess/visualization/material_properties.cc:131]: Ah, that's a left-over from a previous iteration. Let me remove it.

**Comment by bangerth** [comment]
Good spot, @tjhei ! I fixed it!

**Review by tjhei — APPROVED** [review]
Looks good to me!

**Comment by gassmoeller** [comment]
Thanks for adding the tests. Ready to merge.

**Pull request merged**